### PR TITLE
WIP: Revamp Docs: Rewrite Index Page

### PR DESCRIPTION
**Pull request by SanketDG**
### Thanks for contributing, make sure you address all the checklists (for details on how see
[development documentation](https://virtualenv.pypa.io/en/latest/development.html#development))!

This is some preliminary work towards https://github.com/pypa/virtualenv/issues/1801. Since this will completely break the current docs, I am not sure how should I go about this. I would still advise you to consider the workflow I have suggested in https://github.com/pypa/virtualenv/issues/1801#issuecomment-624678532

- [x] ran the linter to address style issues (``tox -e fix_lint``)
- [x] wrote descriptive pull request text
- [x] ensured there are test(s) validating the fix
- [x] added news fragment in ``docs/changelog`` folder
- [x] updated/extended the documentation

### CODE REVIEW

File: docs/index.rst [hunk 1]
```diff
@@ -30,17 +30,15 @@ Virtualenv
...
+a working Python environment that is isolated from the rest of the system.
 
-- is slower (by not having the ``app-data`` seed method),
-- is not as extendable,
-- cannot create virtual environments for arbitrarily installed python versions (and automatically discover these),
-- is not upgrade-able via `pip <https://pip.pypa.io/en/stable/installing/>`_,
-- does not have as rich programmatic API (describe virtual environments without creating them).
+.. note::
+  Since Python ``3.3``, a subset of it has been integrated into the standard library under the `venv module <https://docs.python.org/3/library/venv.html>`_. The
+  ``venv`` module does not offer all features of this library. <Difference link here>
```
Comment (gaborbernat) [inline docs/index.rst:38]: do add the difference link

File: docs/index.rst [hunk 2]
```diff
@@ -30,17 +30,15 @@ Virtualenv
...
 
-The basic problem being addressed is one of dependencies and versions, and indirectly permissions.
+The need for isolation of python environments comes when there are two python
```
Comment (gaborbernat) [inline docs/index.rst:40]: I liked better this before. We're not just talking here about applications. For example when I'm testing a library that's not an application, but I still need an isolated python environment, not?

Comment (SanketDG) [reply docs/index.rst:40]: @gaborbernat Absolutely, I agree. I put in application because that's how it's present in the rest of the paragraph below:
> magine you have an application that needs version ``1`` of ``LibFoo``, but another application requires version
``2``. How can you use both these libraries? If you install everything into your host python (e.g. ``python3.8``)
it's easy to end up in a situation where two packages have conflicting requirements.

Should I rephrase this too?

Comment (SanketDG) [reply docs/index.rst:40]: I rewrote this loosely inspired from https://docs.python.org/3/tutorial/venv.html#introduction

File: docs/index.rst (hunk 1 above)
Comment (SanketDG) [reply docs/index.rst:38]: Done in the latest iteration.

File: docs/index.rst [hunk 3]
```diff
@@ -30,51 +30,64 @@ Virtualenv
...
 
 In all these cases, ``virtualenv`` can help you. It creates an environment that has its own installation directories,
 that doesn't share libraries with other virtualenv environments (and optionally doesn't access the globally installed
 libraries either).
 
-Useful links
-------------
+Quick Start
+-----------
+The recommended way to install virtualenv is through pip: (for other methods see :doc:`Installation <installation>`)
```
Comment (gaborbernat) [inline docs/index.rst:59]: I strongly disagree with this. installing virtualenv with pip causes all kind of issues I'm not comfortable recommending.

File: docs/index.rst [hunk 4]
```diff
@@ -30,51 +30,64 @@ Virtualenv
...
+.. code-block:: console
 
-* :pypi:`virtualenvwrapper` - a useful set of scripts for creating and deleting virtual environments
-* :pypi:`pew` - provides a set of commands to manage multiple virtual environments
-* :pypi:`tox` - a generic virtualenv management and test automation command line tool, driven by a ``tox.ini``
-  configuration file
-* :pypi:`nox` - a tool that automates testing in multiple Python environments, similar to tox,
-  driven by a ``noxfile.py`` configuration file
+    $ Python -m pip install --user virtualenv
+    $ virtualenv --version
```
Comment (gaborbernat) [inline docs/index.rst:64]: Not sure what's the point of this. Also probably ``python -m virtualenv is the sane way to access this`` as on Windows the installed bin folder path might not be on the path.

File: docs/index.rst [hunk 5]
```diff
@@ -30,51 +30,64 @@ Virtualenv
...
 
-**Tutorials**
+To create a virtualenv using the default Python instance available on the system
```
Comment (gaborbernat) [inline docs/index.rst:66]: Well not sure what qualifies default python. It's more like the python under what virtalenv is installed.

File: docs/index.rst [hunk 6]
```diff
@@ -30,51 +30,64 @@ Virtualenv
...
 
-**Presenting how the package works from within**
+    $ virtualenv venv/
 
-* `Bernat Gabor: status quo of virtual environments <https://www.youtube.com/watch?v=o1Vue9CWRxU>`_
-* `Carl Meyer: Reverse-engineering Ian Bicking's brain: inside pip and virtualenv
-  <http://pyvideo.org/video/568/reverse-engineering-ian-bicking--39-s-brain--insi>`_
+To activate a virtualenv
+
+Bash
```
Comment (gaborbernat) [inline docs/index.rst:74]: I'd not enumerate any of these. Makes a lot of assumption that the user's shell is bash, often it's not.

File: docs/index.rst [hunk 7]
```diff
@@ -30,51 +30,64 @@ Virtualenv
...
+.. code-block:: console
+
+    $ source venv/bin/activate
+
+Windows
+^^^^^^^
+
+.. code-block:: console
+
+    venv/bin/activate.bat
```
Comment (gaborbernat) [inline docs/index.rst:86]: Also no for this. Assumes user runs with command prompt, but that's not the case with git bash or PowerShell.

File: docs/index.rst (hunk 3 above)
Comment (SanketDG) [reply docs/index.rst:59]: I will replace with `pipx` then, thanks (y)

Comment (gaborbernat) [reply docs/index.rst:59]: I don't think needs replaced, needs to be removed. Maybe a link to the install section but no need to point out any method in particular. The user might even pull it in from its operating system package manager, it should not matter.

File: docs/index.rst (hunk 6 above)
Comment (SanketDG) [reply docs/index.rst:74]: The idea is to target the most commonly used shells/platforms. But I understand.

A list of exhaustive activators are not present in the documentation.

Do you want me to create that? Is there an alternative way that you have in mind for specifying the activators in a platform agnostic way?

Comment (gaborbernat) [reply docs/index.rst:74]: Mainly because it's not set in stone users might create more via https://virtualenv.pypa.io/en/latest/extend.html#python-discovery. A default shell targeted is documented under https://virtualenv.pypa.io/en/latest/cli_interface.html#section-activators. Not sure what more you would want. https://virtualenv.pypa.io/en/latest/user_guide.html#activators details the contract.

File: docs/index.rst (hunk 3 above)
Comment (SanketDG) [reply docs/index.rst:59]: I wanted to write a "Quick Start" and a "Tutorial" mostly aimed at beginners, who would come to the docs. Since the installation and activation part is best delegated to its original documentation (which is a little verbose IMO), do you want me to remove this section entirely?

Comment (gaborbernat) [reply docs/index.rst:59]: I'd say yes.

Comment (SanketDG) [reply docs/index.rst:59]: Sure, then what would you suggest would be the best way to have a guide/tutorial that is well suited for beginners, and doesn't have to go through pages of documentation to figure out how to use `virtualenv`

Comment (SanketDG) [reply docs/index.rst]: I have removed the Quick Start section

File: docs/index.rst (hunk 6 above)
Comment (SanketDG) [reply docs/index.rst]: I have removed the Quick Start section

File: docs/index.rst (hunk 7 above)
Comment (SanketDG) [reply docs/index.rst]: I have removed the Quick Start section

File: docs/index.rst (hunk 5 above)
Comment (SanketDG) [reply docs/index.rst]: I have removed the Quick Start section

File: docs/index.rst (hunk 4 above)
Comment (SanketDG) [reply docs/index.rst]: I have removed the Quick Start section

File: docs/useful_links.rst [hunk 8]
```diff
@@ -0,0 +1,19 @@
+Useful links
+============
+
+**Related projects, that build abstractions on top of virtualenv**
+
+* `virtualenvwrapper <https://virtualenvwrapper.readthedocs.io/en/latest/>`_ - a useful set of scripts for creating and deleting virtual environments
```
Comment (gaborbernat) [inline docs/useful_links.rst:6]: Please restore these links to pypi. The pypi contains links to the documentation, code base etc.

Comment (gaborbernat) [reply docs/useful_links.rst:6]: Add back pew.

File: docs/useful_links.rst [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  driven by a ``noxfile.py`` configuration file
+
+**Tutorials**
+
+* `Corey Schafer tutorial <https://www.youtube.com/watch?v=N5vscPTWKOk>`_ on how to use it
+* `Using virtualenv with mod_wsgi <http://code.google.com/p/modwsgi/wiki/VirtualEnvironments>`_
+
+**Presenting how the package works from within**
+
+* `Bernat Gabor: status quo of virtual environments <https://www.youtube.com/watch?v=o1Vue9CWRxU>`_
```
Comment (gaborbernat) [inline docs/useful_links.rst:19]: Add back Carl Meyers video; it's an interesting case study on how things were in the past.

File: docs/index.rst [hunk 10]
```diff
@@ -30,51 +30,31 @@ Virtualenv
...
+library. The applications will often require a specific version  of that library where a particular
+bug has been fixed. Imagine you have an application that needs version ``1.2.3`` of ``foo``,
+but another application requires version ``2.3.4``. Installing one version of the library would
+make the other application unable to run. So, If you install everything into your system Python
+it's easy to end up in a situation where two applications have conflicting requirements.
 
 Or more generally, what if you want to install an application *and leave it be*? If an application works, any change
 in its libraries or the versions of those libraries can break the application. Also, what if you can't install packages
-into the global ``site-packages`` directory, due to not having permissions to change the host python environment?
+into the global ``site-packages`` directory, due to not having permissions to change the system Python environment?
```
Comment (gaborbernat) [inline docs/index.rst:51]: The term host is accurate, not the system. Doesn't have to be system python install; it can be side-load installed too.

File: docs/index.rst [hunk 11]
```diff
@@ -30,51 +30,31 @@ Virtualenv
...
-- does not have as rich programmatic API (describe virtual environments without creating them).
+.. note::
+  Since Python ``3.3``, a subset of it has been integrated into the standard library under the `venv module <https://docs.Python.org/3/library/venv.html>`_. The
+  ``venv`` module does not offer all features of this library. See the :doc:`differences <differences_with_venv>` betwwen virtualenv and venv.
 
-The basic problem being addressed is one of dependencies and versions, and indirectly permissions.
-Imagine you have an application that needs version ``1`` of ``LibFoo``, but another application requires version
-``2``. How can you use both these libraries? If you install everything into your host python (e.g. ``python3.8``)
-it's easy to end up in a situation where two packages have conflicting requirements.
+The need for isolation of Python environments comes when there are two Python
```
Comment (gaborbernat) [inline docs/index.rst:40]: I prefer the initial formulation. I find the new sentences hard to follow and inaccurate in some places.

File: docs/differences_with_venv.rst [hunk 12]
```diff
@@ -0,0 +1,17 @@
+Differences with venv
+=====================
+
+- ``venv`` is bundled with Python itself (starting from 3.3+), ``virtualenv``
+  is a third party package. This allows for ``virtualenv`` to have greater
+  flexibility when dealing with several Python versions.
```
Comment (gaborbernat) [inline docs/differences_with_venv.rst:6]: I don't think the flexibility comes from the fact that it's bundled.

File: docs/differences_with_venv.rst [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+- ``virtualenv`` allows further :doc:`extendibility <extend>` to override default
+  functionality through a plugin system.
+- ``virtualenv`` can create virtual environments through its interpreter discovery
+  technique from arbitrary Python locations, they do not be installed or be on the
```
Comment (gaborbernat) [inline docs/differences_with_venv.rst:10]: This is not true.

File: docs/differences_with_venv.rst [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+  ``PATH``.
+- ``venv`` can only create virtual environments for the Python version it is
+  installed with, ``virtualenv`` supports both Python 2 and 3 and also other flavours
+  like ``CPython`` and ``PyPy``.
+- ``virtualenv`` has a rich programmatic API (describe virtual environments without
+  creating them).
+- ``virtualenv`` has better support across different platforms and shells.
```
Comment (gaborbernat) [inline docs/differences_with_venv.rst:17]: This is not true at platform level; shells maybe, but that is not the point here. venv adds support for these extra shells... but they take time to ship.

File: docs/differences_with_venv.rst [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+- ``venv`` is bundled with Python itself (starting from 3.3+), ``virtualenv``
+  is a third party package. This allows for ``virtualenv`` to have greater
+  flexibility when dealing with several Python versions.
+- ``virtualenv`` allows further :doc:`extendibility <extend>` to override default
+  functionality through a plugin system.
+- ``virtualenv`` can create virtual environments through its interpreter discovery
+  technique from arbitrary Python locations, they do not be installed or be on the
+  ``PATH``.
+- ``venv`` can only create virtual environments for the Python version it is
+  installed with, ``virtualenv`` supports both Python 2 and 3 and also other flavours
```
Comment (gaborbernat) [inline docs/differences_with_venv.rst:13]: You mention other flavours, but then you point out two we support. The other implies you already mentioned one and you're including new ones.

File: docs/differences_with_venv.rst [hunk 16]
```diff
@@ -0,0 +1,17 @@
+Differences with venv
+=====================
+
+- ``venv`` is bundled with Python itself (starting from 3.3+), ``virtualenv``
```
Comment (gaborbernat) [inline docs/differences_with_venv.rst:4]: You removed the performance entry which is the main reason one would pick this.

File: docs/useful_links.rst (hunk 8 above)
Comment (SanketDG) [reply docs/useful_links.rst:6]: I had initially written the reasonings for this in my initial spec. Apologies, I thought you had approved those.

> Please restore these links to pypi. The pypi contains links to the documentation, code base etc.

PyPI is not an interface that a lot of people visit everyday.

> Add back pew.

`pew`'s last release/commit is from 2019. How is it relevant?

File: docs/useful_links.rst (hunk 9 above)
Comment (SanketDG) [reply docs/useful_links.rst:19]: Given the video's quality and age, I am much more inclined to write a document how things were before in the past.

File: docs/differences_with_venv.rst (hunk 14 above)
Comment (SanketDG) [reply docs/differences_with_venv.rst:17]: This comes naively from your presentation.

https://gaborbernat.github.io/euro_python_2019/#/

File: docs/differences_with_venv.rst (hunk 13 above)
Comment (SanketDG) [reply docs/differences_with_venv.rst:10]: This also comes naively from your presentation https://gaborbernat.github.io/euro_python_2019/#/

But I could be wrong on the "installation" part.

File: docs/differences_with_venv.rst (hunk 16 above)
Comment (SanketDG) [reply docs/differences_with_venv.rst:4]: It defines no reason for said performance.

And that performance only comes into play if someone uses the `app-seed` feature, right?

File: docs/differences_with_venv.rst (hunk 12 above)
Comment (SanketDG) [reply docs/differences_with_venv.rst:6]: I don't understand, it's not bundled, that's why it has greater flexibility?

I also deduced this from your resentation, given there are multiple installation options and support for any Python version.

File: docs/differences_with_venv.rst (hunk 15 above)
Comment (SanketDG) [reply docs/differences_with_venv.rst:13]: I will rephrase this.

File: docs/index.rst (hunk 11 above)
Comment (SanketDG) [reply docs/index.rst:40]: This is adapted from https://docs.python.org/3/library/venv.html, and it seems readable to me.

File: docs/index.rst (hunk 10 above)
Comment (SanketDG) [reply docs/index.rst:51]: `host` doesn't mean intuitive at all, and so doesn't `system`.

I will revert the change.

File: docs/useful_links.rst (hunk 8 above)
Comment (SanketDG) [reply docs/useful_links.rst:6]: I had described this in my initial spec on the revamp issue

> Please restore these links to pypi. The pypi contains links to the documentation, code base etc.

PyPI isn't a interface that is visited a lot, I am not sure why anything else over the official docs is preferred here



> Add back pew.

Pew's last commit/release was in 2019, I don't see how it's relevant. There are probably a bunch of other tools that are used in the ecosystem today that deserves a name here

File: docs/useful_links.rst (hunk 9 above)
Comment (SanketDG) [reply docs/useful_links.rst:19]: Given the quality and the age of the video, I'd be inclined to write a document on how things were in the past.

**Comment by SanketDG** [comment]
@gaborbernat 

Apologies if this has been less convincing. I have replied to all the comments above.

Most of the points here were adapted from your talk in EuroPython, but I understand me being not familiar with the system has caused a lot of changes to not make sense.

I will close this PR if these changes do not align with your vision and thought process.

File: docs/useful_links.rst (hunk 8 above)
Comment (gaborbernat) [reply docs/useful_links.rst:6]: Not everything needs a new release to be relevant. pew is even more popular than nox, so should we remove that too; see https://pepy.tech/project/nox vs https://pepy.tech/project/pew

File: docs/index.rst (hunk 11 above)
Comment (gaborbernat) [reply docs/index.rst:40]: We'll have to agree to disagree on this then 😊

File: docs/differences_with_venv.rst (hunk 16 above)
Comment (gaborbernat) [reply docs/differences_with_venv.rst:4]: It's what most visible to end-users though and is on by default.

**Comment by gaborbernat** [comment]
@SanketDG maybe this type of changes are not effective via PR. Feels this could be fleshed out much quicker in a sprint (face-to-face discussion). This way feels arbitrary what gets removed, where and how; so we're going to have a lot of back and forth on it; as IMHO some of them are a matter of taste. 

> Let's step back and understand what you try to achieve here? Is it to make it more accessible to someone totally new to Python? If so maybe we just need to add a single page that addresses the need. Something like `High level introduction` and mention on the front page if you want a high level why and how to use, read this page. For who is interested in the details read other sections.

Still waiting for an answer on this 👍

**Comment by SanketDG** [comment]
> Still waiting for an answer on this

There definitely seems to be a place for this.

How do you want me to go about it?

**Comment by gaborbernat** [comment]
Instead of trying to rewrite parts, just add a new beginner friendly page?

**Comment by SanketDG** [comment]
> Instead of trying to rewrite parts, just add a new beginner friendly page?

I can do that, but as you said a little bit of back and forth discussion would help here (would gitter be the right thing here?)

Is there anything relevant in this PR that you wish to keep? I will otherwise close it then.

**Comment by gaborbernat** [comment]
I think let's close it.

**Comment by gaborbernat** [comment]
> I can do that, but as you said a little bit of back and forth discussion would help here (would gitter be the right thing for this?)

If you keep it as an addition of new document rather than changing existing this might be less disruptive, and you can start with a draft proposal.
